pkg/inframetadata/reporter.go

```diff
@@ -0,0 +1,149 @@
+	"github.com/DataDog/opentelemetry-mapping-go/pkg/otlp/attributes"
+	"github.com/DataDog/opentelemetry-mapping-go/pkg/otlp/attributes/source"
+)
+
+const (
+	// AttributeDatadogHostUseAsMetadata states if a payload should be used for host metadata.
+	// It overrides the default behavior (see below).
+	AttributeDatadogHostUseAsMetadata = "datadog.host.use_as_metadata"
+
+	// shouldUseByDefault specifies if the payloads should be used by default when the
+	// use_as_metadata resource attribute is missing.
+	shouldUseByDefault = false
+)
+
+type Pusher interface {
+	// Push host metadata to a remote endpoint.
+	// MUST be safe to call concurrently.
+	Push(context.Context, payload.HostMetadata) error
+}
+
+// Reporter of host metadata based on pcommon.Resource payloads.
+type Reporter struct {
+	// logger (sampled) for warnings.
+	logger *zap.Logger
+	// hostMap storing the host metadata.
+	hostMap *hostmap.HostMap
+	// pusher of host metadata.
+	pusher Pusher
+	// closeCh can stop the host metadata reporting.
+	closeCh chan struct{}
+	// ticker for periodic host metadata reporting.
+	ticker *time.Ticker
+}
+
+// Copied over from github.com/open-telemetry/opentelemetry-collector/blob/14c039d/exporter/exporterhelper/queued_retry.go#L269
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+func createSampledLogger(logger *zap.Logger) *zap.Logger {
+	if logger.Core().Enabled(zapcore.DebugLevel) {
+		// Debugging is enabled. Don't do any sampling.
+		return logger
+	}
+
+	// Sample all messages to 1 per 10 seconds initially,
+	// and 1/100 of messages after that.
+	opts := zap.WrapCore(func(core zapcore.Core) zapcore.Core {
+		return zapcore.NewSamplerWithOptions(core, 10*time.Second, 1, 100)
+	})
+	return logger.WithOptions(opts)
+}
+
+// NewReporter creates a new host metadata reporter.
+// The reporter consumes pcommon.Resources through its 'Consume' method and merges them into payload.HostMetadata payloads.
+// It then periodically exporters these functions through the pusher with a specified period.
+func NewReporter(logger *zap.Logger, pusher Pusher, period time.Duration) (*Reporter, error) {
+	hostMap, err := hostmap.New()
+	if err != nil {
+		return nil, fmt.Errorf("failed to build host map: %w", err)
+	}
+
+	return &Reporter{
+		logger:  createSampledLogger(logger),
+		hostMap: hostMap,
+		pusher:  pusher,
+		closeCh: make(chan struct{}),
+		ticker:  time.NewTicker(period),
+	}, nil
+}
+
+// checkResource to see if it should be used by default.
+// A resource is usable if 'AttributeDatadogHostUseAsMetadata' is true or shouldUseByDefault is true.
+func checkResource(res pcommon.Resource) (bool, error) {
+	shouldUse := shouldUseByDefault
+	if val, ok := res.Attributes().Get(AttributeDatadogHostUseAsMetadata); ok {
+		if val.Type() != pcommon.ValueTypeBool {
+			return false, fmt.Errorf("%q has type %q, expected \"Bool\"", AttributeDatadogHostUseAsMetadata, val.Type())
```

Comment:
I'm wondering if we can simplify the code here and simply have a fall back where we take the string representation of the attribute (when it's not bool) and if it's [parsable](https://pkg.go.dev/strconv#ParseBool) use that. If there is an error, we could simply log it here.

We could then get rid of the returned error value here, and also in `Consume`. Is it important for consume to return an error in this case? Isn't logging enough? That's what happens ultimately with the error anyway isn't it?

Reply:
> I'm wondering if we can simplify the code here and simply have a fall back where we take the string representation of the attribute (when it's not bool) and if it's [parsable](https://pkg.go.dev/strconv#ParseBool) use that. If there is an error, we could simply log it here.

I chose a more conservative approach since it's not clear to me what the behavior should be with a string. Should we accept `1` as true? What about `True`? `TRUE`? `yes`? Is it going to be confusing to the user if we don't accept any of those?

> We could then get rid of the returned error value here, and also in Consume. Is it important for consume to return an error in this case? Isn't logging enough? That's what happens ultimately with the error anyway isn't it?

Logging is what ultimately will happen in this case, yes, but it's not necessarily the case if we reuse this code elsewhere (e.g. in the backend). I think returning an error makes it more general.

Reply:
I feel like we don't have to take this decision now, expanding the set of values we support is a backwards compatible change and we can do it once we have had some users try this out. Put it simply: if we change it now, we commit to supporting the whole set of values forever, if we don't we can always change it later.

Do you think this is important to support now (e.g. do you feel like this would be confusing as is for the user)?